src/magicport/dprinter.d

```diff
@@ -0,0 +1,1719 @@
+        foreach(i; 0..ast.types.length)
+        {
+            if (i)
+                print(", ");
+            else
+            {
+                visitX(ast.stc | STCvirtual);
+                visitX(ast.types[i]);
+                print(" ");
+            }
+            visitIdent(ast.ids[i]);
+            if (ast.inits[i])
+            {
+                print(" = ");
+                this.inittype = ast.types[i];
+                visitX(ast.inits[i]);
+                inittype = null;
+            }
+        }
+    }
+
+    bool isClass(Type t)
+    {
+        if (auto ct = cast(ClassType)t)
+        {
+            auto id = ct.id;
+            if (id.startsWith("class "))
+                id = id[6..$];
+            if (classTypes.lookup(id))
+                return true;
+            foreach(sd; scan.structsUsingInheritance)
+            {
+                if (sd.id == id)
+                {
+                    return true;
+                }
+            }
+            if (rootClasses.lookup(id))
+                return true;
+        }
+        return false;
+    }
+
+    override void visit(ConstructDeclaration ast)
+    {
+        if (isClass(ast.type))
+        {
+            stackclasses ~= ast.id;
+            print("scope ");
+            visitX(new PointerType(ast.type));
+            print(" ");
+            visitIdent(ast.id);
+            print(" = new ");
+            visitX(ast.type);
+            print("(");
+            printArgs(ast.args);
+            print(")");
+        }
+        else
+        {
+            print("auto ");
+            visitIdent(ast.id);
+            print(" = ");
+            visitX(ast.type);
+            print("(");
+            printArgs(ast.args);
+            print(")");
+        }
+    }
+
+    void versionCommon(T)(T ast)
+    {
+        foreach(i, c; ast.cond)
+        {
+            if (c)
+            {
+                if (i)
+                    print("else ");
+                auto ie = cast(IdentExpr)c;
+                auto le = cast(LitExpr)c;
+                if (ie)
+                {
+                    switch(ie.id)
+                    {
+                    case "DEBUG":          println("debug"); break;
+                    case "UNITTEST":       println("version (unittest)"); break;
+
+                    case "_WIN32":         println("version (Windows)"); break;
+                    case "POSIX":          println("version (Posix)"); break;
+                    case "__linux__":      println("version (linux)"); break;
+                    case "__APPLE__":
+                    case "MACINTOSH":      println("version (OSX)"); break;
+                    case "__FreeBSD__":    println("version (FreeBSD)"); break;
+                    case "__OpenBSD__":    println("version (OpenBSD)"); break;
+                    case "__sun":          println("version (Solaris)"); break;
+
+                    case "IN_GCC":         println("version (GNU)"); break;
```

Comment:
I would argue that ``IN_GCC`` is not quite the same as ``version (GNU)`` - this assumption would mean that you can't use DMD as a host D compiler to build GDC.